### DISCUSSION TITLE
Tracing for Production Applications

### DISCUSSION THREAD

**Discussion by kevin-m-kent** [microsoft/promptflow#2594] [Q&A]
Hi - I see there is useful output after each run tracing the activity of each node (inputs, outputs, execution time, etc) when I run the flow locally under .promptflow/chat.detail.json. I am interested in sending these traces to a place like Application Insights for my promptflow instance that I have deployed as an Azure web application. I see in the tracing and dev modules there are references to application insights but I can't quite figure out how one would go about sending these traces to a place like application insights. Are there any examples of this or guidance? Thank you.

**Comment by kevin-m-kent** [comment]
Is it as simple as setting APPLICATIONINSIGHTS_CONNECTION_STRING with devkit installed?

**Reply by wxpjimmy** [reply]
Hi @kevin-m-kent,  if you just want to collect traces to app insight for your deployed app, yes, you can configure this "APPLICATIONINSIGHTS_CONNECTION_STRING" in your app configuration and you don't need to install the devkit in the deployed app.

**Reply by kevin-m-kent** [reply]
ok thanks! I did this but I am not getting any traces to my instance. Should this work out of the box? Any particular promptflow version? (I'm on 1.5.0 at the moment).

**Reply by kevin-m-kent** [reply]
upgraded but I think I figured out why it isn't working, some issue with None values:

``` 
failed to record metrics: object of type 'NoneType' has no len()
```

seems it might be failing on this line - https://github.com/microsoft/promptflow/blob/6b9c5c6f1d62da1a7b9d307891902c6fac8af0c1/src/promptflow-core/promptflow/core/_serving/monitor/metrics.py#L216

maybe I should migrate this to an issue? Looking at the debug output it does look like some of my nodes have None for system_metrics.

**Reply by kevin-m-kent** [reply]
Opened https://github.com/microsoft/promptflow/issues/2644

**Reply by wxpjimmy** [reply]
@kevin-m-kent trace and metrics are two different things, the metrics issue won't impact your collected trace data, have you checked your trace data after upgrading to the latest PF package?

**Reply by kevin-m-kent** [reply]
yes I did and I wasn't getting any traces. And the error message I was getting was coming from the function I listed above. there is no dependency between record_tracing_metrics failing and the trace collection and upload path?
